File: java/parser/src/main/java/gov/lbl/parser/parser/DeclarationVisitor.java
package gov.lbl.parser.parser;

import gov.lbl.antlr4.visitor.modelicaBaseVisitor;
import gov.lbl.antlr4.visitor.modelicaParser;
import gov.lbl.parser.domain.Array_subscripts;
import gov.lbl.parser.domain.Declaration;
import gov.lbl.parser.domain.Modification;

public class DeclarationVisitor extends modelicaBaseVisitor<Declaration> {
    @Override
    public Declaration visitDeclaration(modelicaParser.DeclarationContext ctx) {
        String identifier = ctx.IDENT().getText();
        
        Array_subscriptsVisitor array_subscriptsVisitor = new Array_subscriptsVisitor();
        Array_subscripts array_subscripts = ctx.array_subscripts() == null ? null : ctx.array_subscripts().accept(array_subscriptsVisitor);
        
        ModificationVisitor modificationVisitor = new ModificationVisitor();
        Modification modification = ctx.modification() == null ? null : ctx.modification().accept(modificationVisitor);        
        
        return new Declaration(identifier, array_subscripts, modification);
    }
}

File: jsParser/parser/DeclarationVisitor.js
const modelicaVisitor = require('../antlrFiles/modelicaVisitor').modelicaVisitor
const Declaration = require('../domain/Declaration')

const Array_subscriptsVisitor = require('./Array_subscriptsVisitor')
const ModificationVisitor = require('./ModificationVisitor')

class DeclarationVisitor {
  constructor () {
    modelicaVisitor.call(this)
    return this
  }

  visitDeclaration (ctx) {
    const identifier = ctx.IDENT() ? ctx.IDENT().getText() : ''
    let array_subscripts = null
    let modification = null

    if (ctx.array_subscripts()) {
      const array_subscriptsVisitor = new Array_subscriptsVisitor.Array_subscriptsVisitor()
      array_subscripts = array_subscriptsVisitor.visitArray_subscripts(ctx.array_subscripts())
    }
    if (ctx.modification()) {
      const modificationVisitor = new ModificationVisitor.ModificationVisitor()
      modification = modificationVisitor.visitModification(ctx.modification())
    }

    return new Declaration.Declaration(identifier, array_subscripts, modification)
  }
};

DeclarationVisitor.prototype = Object.create(modelicaVisitor.prototype)

exports.visitDeclaration = this.visitDeclaration
exports.DeclarationVisitor = DeclarationVisitor
